remote-trait-object/src/port.rs

```diff
@@ -19,42 +19,83 @@ mod server;
 mod types;
 
 pub use self::types::Handler;
+use crate::forwarder::ServiceForwarder;
 use crate::ipc::multiplex::{MultiplexResult, Multiplexer};
 use crate::ipc::{IpcRecv, IpcSend};
+use crate::service::*;
+use client::Client;
+use server::Server;
+use std::sync::{Arc, Weak};
+
+pub trait Port: Send + Sync + 'static {
+    fn call(&self, arg: String) -> String;
+    fn delete_request(&self, id: ServiceObjectId);
+    fn register(&self, handle_to_register: Box<dyn Service>) -> ServiceObjectId;
+}
 
 pub struct BasicPort {
-    multiplexer: Option<Multiplexer>,
-    server: Option<server::Server>,
-    client: client::Client,
+    registry: Arc<ServiceForwarder>,
+    client: Client,
+}
+
+impl Port for BasicPort {
+    fn call(&self, arg: String) -> String {
+        self.client.call(arg)
+    }
+
+    fn delete_request(&self, _id: ServiceObjectId) {
+        unimplemented!()
+    }
+
+    fn register(&self, service: Box<dyn Service>) -> ServiceObjectId {
+        // FIXME
+        self.registry.add_service("DUMMY".to_owned(), service);
+        ServiceObjectId {
+            index: 1234,
+        }
+    }
 }
 
 impl BasicPort {
-    pub fn new<H, IpcSender, IpcReceiver>(ipc_send: IpcSender, ipc_recv: IpcReceiver, handler: H) -> Self
-    where
-        H: Handler + Send + 'static,
-        IpcSender: IpcSend + 'static,
-        IpcReceiver: IpcRecv + 'static, {
+    pub fn new(client: Client) -> Self {
+        Self {
+            registry: Arc::new(ServiceForwarder::new()),
+            client,
+        }
+    }
+}
+
+pub struct PortInstance {
+    multiplexer: Option<Multiplexer>,
+    server: Option<Server>,
+    port: Arc<BasicPort>,
+}
+
+impl PortInstance {
+    pub fn new<S: IpcSend + 'static, R: IpcRecv + 'static>(ipc_send: S, ipc_recv: R) -> Self {
         let MultiplexResult {
             multiplexer,
             request_recv,
             response_recv,
             multiplexed_send,
         } = Multiplexer::multiplex(ipc_send, ipc_recv);
         let client = client::Client::new(multiplexed_send.clone(), response_recv);
-        let server = server::Server::new(handler, multiplexed_send, request_recv);
-        Self {
-            client,
-            server: Some(server),
+        let port = Arc::new(BasicPort::new(client));
+        let server = server::Server::new(port.registry.clone(), multiplexed_send, request_recv);
+
+        PortInstance {
             multiplexer: Some(multiplexer),
+            server: Some(server),
+            port,
         }
     }
 
-    pub fn call(&self, message: String) -> String {
-        self.client.call(message)
+    pub fn get_port(&self) -> Weak<dyn Port> {
```

Comment:
PortInstance::get_port seems strange. We are getting a port from another port.
Currently, I don't have a good name. Let's change the name when we find a good name.

Reply:
@majecty Instead of changing this, `PortInstance` -> `Instance` might work. It's true that a single instance corresponds to a single instance of `remote-trait-object` context, but the user doesn't actually need to know about the concept `Port`

Reply:
Using Instance looks good. Then let's create a new file for the `Instance`.